### ISSUE TITLE
Tree Structure table export xlsx includes only parent data

### ISSUE THREAD

**Issue by SultanaZinia** [olifolkerd/tabulator#2095]
I have used tabulator for the tree structure table but while I am exporting xlsx of that table it only includes the parent data. I want the export with child data also.

**Comment by aklaver** [comment]
The child data are not rows in their right, more like properties of their parent. To get at them you are going to need to create a custom file formatter:
http://tabulator.info/docs/4.2/download#custom
In that grab the _children array and loop over that to get the data into a row format. This is high level view, if I get some time I will see if I can come up with something more specific.

**Comment by SultanaZinia** [comment]
Thank You for the answer. I solved this problem. but I didn't use custom file formattor rather I changed in the tablulator.js file where the download functions are available. so far, It's working.

**Comment by aklaver** [comment]
Could you show what you did and/or submit a pull request?

**Comment by olifolkerd** [comment]
I am adding tree children to the up comming 4.3 release (that is download, clipboard, print, html etc)

As a general note, i had a busy month at works so im running a bit behind, but it should be out by the end of June.

Cheers

Oli :)

**Comment by ZhanbotaSA** [comment]
> Thank You for the answer. I solved this problem. but I didn't use custom file formattor rather I changed in the tablulator.js file where the download functions are available. so far, It's working.

Can you please explain where and how to change code in tabulator.js for download _children data? I can't resolve this problem( 
~Thanks

**Comment by britvik** [comment]
@olifolkerd Hi, I tried to download data to xls but the tree children were not included. Is it really included in the 4.3 version? I couldn't find a commit that would add such feature, nor saw anything while debugging the download. Luckily I found out there is a `downloadDataFormatter` and I was able to flatten the data and now it is included in xls :).

Here is the code if anybody needs it (or you want to add it to the tabulator):
```
// this is an option passed to Tabulator constructor
downloadDataFormatter: (dataAndCalcs) => ({
  data: flatten((dataAndCalcs as any).data, this.config.dataTreeChildField),
  calcs: (dataAndCalcs as any).calcs,
}),

// helper function
function flatten(arr: any[], childrenAttrName = '_children') {
  return arr ? arr.reduce((result, item) => [
    ...result,
    item,
    ...flatten(item[childrenAttrName])
  ], []) : []
}
```

Also the argument passed to `downloadDataFormatter` has a wrong type in typescript. It's `any[]` but tabulator is passing an object containing `calcs` and `data`  - therefore the `as any` cast :-(

**Comment by SultanaZinia** [comment]
> > Thank You for the answer. I solved this problem. but I didn't use custom file formattor rather I changed in the tablulator.js file where the download functions are available. so far, It's working.
> 
> Can you please explain where and how to change code in tabulator.js for download _children data? I can't resolve this problem(
> ~Thanks

@ZhanbotaSA Hi, Sorry for my super late answer. I used a recursive function to parse the child data. there is a function named `parseRows ` which is defined in many places to retrieve  the data. I slightly changed the function like this and included this same changes in all the download types like :
`
function parseRows(data) {
				//build table rows
				data.forEach(function (row) {
					//body.push(parseRow(row));
					getChildData(row);
				});
			}
// helper recursive function to fetch child data
function getChildData(parent) {
				var parent_data = Object.assign({}, parent);
				delete parent_data["_children"];
				body.push(parseRow(parent_data));
				if ("_children" in parent) {
					for (var i = 0, l = parent["_children"].length; i < l; ++i) {
					var child = parent["_children"][i];
					getChildData(child);
					}
				}   
			}
`
